File: merida.js
class Enemy {
  constructor(x, y, health = 300) {
    this.x = x;
    this.y = y;
    this.health = health;
  }

  damage(amount) {
    this.health -= amount;
  }

  isDead() {
    return this.health <= 0;
  }
}

class Archer {
  constructor(arrows, field, damage = 40) {
    this.x = 0; // starts at the leftmost square
    this.kills = 0;

    this.arrows = arrows;
    this.field = field;
    this.damage = damage;
  }

  moveLeft() {
    if (this.x > 0) {
      this.x--;
    }
  }

  moveRight() {
    if (this.x < this.field.numCols - 1) {
      this.x++;
    }
  }

  shoot() {
    if (this.outOfArrows()) return;

    const enemiesInColumn = this.field.getEnemiesAtColumn(this.x);

    for (let i = 0; i < enemiesInColumn.length; i++) {
      const totalDamage = this.calculateTotalDamage(i);
      let enemy = enemiesInColumn[i];
      enemy.damage(totalDamage);
      if (enemy.isDead()) {
        this.field.vacateTile(enemy.x, enemy.y);
        this.kills++;
      }
    }

    this.arrows--;
  }

  calculateTotalDamage(length) {
    let totalDamage = this.damage;

    for (let i = 0; i < length; i++) {
      totalDamage -= totalDamage * 0.2;
    }
    console.log(totalDamage);
    return Math.floor(totalDamage);
  }

  outOfArrows() {
    return this.arrows <= 0;
  }
}

class Field {
  constructor(rows = 11, cols = 10) {
    this.grid = [];
    this.numRows = rows;
    this.numCols = cols;

    for (let x = 0; x < this.numCols; x++) {
      const column = [];
      for (let y = 0; y < this.numRows; y++) {
        column.push(null);
      }
      this.grid.push(column);
    }
  }

  static loadFromFileContents(contents) {
    let results = contents.split("/");
    if (!results || results.length < 1) return null;
    let rowsCols = results[0].split("x").map(Number);
    const field = new Field(rowsCols[1], rowsCols[0]);
    let coordinatesStr = results[1];
    // Removing the leading and trailing whitespace and parentheses
    coordinatesStr = coordinatesStr.trim().slice(1, -1);

    // Splitting the coordinates string and converting to subarrays
    coordinatesStr.split("), (").map((pair) => {
      const [x, y] = pair.split(",").map(Number);
      field.placeEnemy(x, y);
    });

    return field;
  }

  placeEnemy(x, y) {
    this.grid[x][y] = new Enemy(x, y);
  }

  vacateTile(x, y) {
    this.grid[x][y] = null;
  }

  getEnemiesAtColumn(column) {
    const enemiesInColumn = [];

    for (let y = 0; y < this.numRows; y++) {
      const enemy = this.grid[column][y];
      if (enemy) {
        enemiesInColumn.push(enemy);
      }
    }

    return enemiesInColumn;
  }

  damageEnemiesAtColumn(column, initialDamage) {
    const enemiesInColumn = this.getEnemiesAtColumn(column);

    for (let enemy of enemiesInColumn) {
      enemy.damage(initialDamage);
      if (enemy.isDead()) {
        this.vacateTile(enemy.x, enemy.y);
      }
    }
  }
}

// Example usage:
const fieldContent =
  "10 x 11 / (5, 10), (5, 7), (5, 6), (2, 5), (4, 5), (5, 2)";
const field = Field.loadFromFileContents(fieldContent);
const archer = new Archer(10, field);

console.log(field.grid);
// archer.shoot();
archer.moveRight();
archer.moveRight();
// archer.moveRight();
// archer.moveRight();
// archer.moveRight();
archer.shoot();
// console.log(archer.kills); // Output: 2

File: merida.py
class Enemy:
    def __init__(self, x, y, health=300):
        self.x = x
        self.y = y
        self.health = health

    def damage(self, amount):
        self.health -= amount
      

    def isdead(self):
         return self.health <= 0
        


class Archer:
    def __init__(self, arrows, field, damage=40):
        self.x = 0 # starts at the leftmost square
        self.kills = 0
        self.arrows = arrows
        self.field = field
        self.damage = damage

    def moveleft(self):
        if self.x > 0:
            self.x -= 1

    def moveright(self):
        if self.x < self.field.numCols - 1:
            self.x += 1

    def shoot(self):
        if self.outOfArrows():
            return
    
        enemies_in_column = self.field.getEnemiesAtColumn(self.x)
        for i,enemy in enumerate(enemies_in_column):
            total_damage = self.calculateTotalDamage(i)
            enemy.damage(total_damage)
            if enemy.isdead():
                self.field.vacateTile(enemy.x, enemy.y)
                self.kills += 1
    
        self.arrows -= 1
    def calculateTotalDamage(self,length):
        total_damage = self.damage
      
        for i in range(0, length):
            total_damage -= total_damage * 0.2  # 20% damage loss per enemy pierced

        return int(total_damage)    
 
    def outOfArrows(self):
        return self.arrows <= 0


class Field:
    def __init__(self, rows=11, cols=10):
        self.grid = []
        self.numRows = rows
        self.numCols = cols
        for x in range(self.numCols):
            column = []
            for j in range(self.numRows):
                column.append(None)
            self.grid.append(column)

    @staticmethod
    def loadFromFileContents(contents):
        lines = contents.split("/")
        dimensions = lines[0].strip().split("x")
        rows = int(dimensions[1])
        cols = int(dimensions[0])
        field = Field(rows, cols)

        for line in lines[1:]:
            x, y = line.strip()[1:-1].split(",")
            x = int(x.strip())
            y = int(y.strip())
            field.placeEnemy(x, y)

        return field

    def placeEnemy(self, x, y):
        self.grid[x][y] = Enemy(x, y)

    def vacateTile(self, x, y):
        # opposite of placeEnemy(), just marks a tile as null
        self.grid[x][y] = None

    def getEnemiesAtColumn(self, column):
        # gets all "alive" enemies in the given column number
        enemies_in_column = []

        for y in range(self.numRows):
            enemy = self.grid[column][y]
            if enemy:
                enemies_in_column.append(enemy)

        return enemies_in_column

    def damageEnemiesAtColumn(self, column, initialDamage):
        enemies_in_column = self.getEnemiesAtColumn(column)

        for enemy in enemies_in_column:
            enemy.damage(initialDamage)
            if enemy.isdead():
                self.vacateTile(enemy.x, enemy.y)